pkg/network/tracer/connection/failure/failed_conn_consumer.go

```diff
@@ -0,0 +1,121 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+// Package failure contains logic specific to TCP failed connection handling
+package failure
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const failedConnConsumerModuleName = "network_tracer__ebpf"
+
+// Telemetry
+var failedConnConsumerTelemetry = struct {
+	eventsReceived telemetry.Counter
+	eventsLost     telemetry.Counter
+}{
+	telemetry.NewCounter(failedConnConsumerModuleName, "failed_conn_polling_received", []string{}, "Counter measuring the number of failed connections received"),
+	telemetry.NewCounter(failedConnConsumerModuleName, "failed_conn_polling_lost", []string{}, "Counter measuring the number of failed connections lost (were transmitted from ebpf but never received)"),
+}
+
+// TCPFailedConnConsumer consumes failed connection events from the kernel
+type TCPFailedConnConsumer struct {
+	eventHandler ddebpf.EventHandler
+	once         sync.Once
+	closed       chan struct{}
+	FailedConns  *FailedConns
+}
+
+// NewFailedConnConsumer creates a new TCPFailedConnConsumer
+func NewFailedConnConsumer(eventHandler ddebpf.EventHandler) *TCPFailedConnConsumer {
+	return &TCPFailedConnConsumer{
+		eventHandler: eventHandler,
+		closed:       make(chan struct{}),
+		FailedConns:  NewFailedConns(),
+	}
+}
+
+// Stop stops the consumer
+func (c *TCPFailedConnConsumer) Stop() {
+	if c == nil {
+		return
+	}
+	c.eventHandler.Stop()
+	c.once.Do(func() {
+		close(c.closed)
+	})
+}
+
+func (c *TCPFailedConnConsumer) extractConn(data []byte) {
+	c.FailedConns.Lock()
+	defer c.FailedConns.Unlock()
+	failedConn := (*netebpf.FailedConn)(unsafe.Pointer(&data[0]))
+	failedConnConsumerTelemetry.eventsReceived.Inc()
+
+	// Ignore failed connections that are not in the allow list
+	if _, exists := allowListErrs[failedConn.Reason]; !exists {
+		return
+	}
+	stats, ok := c.FailedConns.FailedConnMap[failedConn.Tup]
```

Comment:
This breaks the FailedConns abstraction a bit. You could add a method to add/update the entry to FailedConns.